### PR TITLE
Task coordinator core logic [2]

### PR DESCRIPTION
**Pull request by AfanasievYuriy**

### CODE REVIEW

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java [hunk 1]
```diff
@@ -1,17 +1,34 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+
 @Getter
 @Setter
 @EqualsAndHashCode
 @AllArgsConstructor
 public class OrderSubmissionDTO {
 
     int orderId;
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java]: modifiers

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    @Transactional
+    public LaundryStateDTO updateLaundryStateWithOrderSubmitted(OrderSubmittedDTO orderSubmittedDTO) {
+        Objects.requireNonNull(orderSubmittedDTO);
+
+        LaundryStateEntity laundryStateEntity = laundryStateRepository.findById(orderSubmittedDTO.getLaundryState().getLaundryId())
+                .orElseThrow(() -> new IllegalArgumentException("LaundryState with id " + orderSubmittedDTO.getOrderId() + "doesn't exist"));
+        OrderDTO order = orderService.getOrderById(orderSubmittedDTO.getOrderId());
+
+        if (laundryStateEntity.getVersion() < orderSubmittedDTO.getLaundryState().getVersion()) {
+            laundryStateEntity.setReservedTime(laundryStateEntity.getReservedTime() - order.getDuration());
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: reserved time should be changed in any case

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+    @Override
+    @Transactional
+    public LaundryStateDTO updateLaundryStateWithOrderProcessed(OrderProcessedDTO orderProcessedDTO) {
+        Objects.requireNonNull(orderProcessedDTO);
+
+        LaundryStateEntity laundryStateEntity = laundryStateRepository.findById(orderProcessedDTO.getLaundryState().getLaundryId())
+                .orElseThrow(() -> new IllegalArgumentException("LaundryState with id " + orderProcessedDTO.getOrderId() + "does not exist"));
+        OrderDTO order = orderService.getOrderById(orderProcessedDTO.getOrderId());
+
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java]: 2 blank lines

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        orderService.updateOrder(order);
+        laundryStateRepository.save(laundryStateEntity);
+
+        return new LaundryStateDTO(laundryStateEntity);
+    }
+
+    @Override
+    @Transactional
+    public LaundryStateEntity getLeastLoadedLaundry() {
+        return laundryStateRepository.getLeastLoadedLaundry();
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/LaundryStateServiceImpl.java:85]: why you created separate method?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java [hunk 5]
```diff
@@ -1,19 +1,80 @@
...
+        existingOrder.setCompletionTime(orderDTO.getCompletionTime());
+        OrderEntity orderSaved = orderRepository.save(existingOrder);
+
+        return new OrderDTO(orderSaved);
+    }
+
+    @Override
+    @Transactional
+    public OrderDTO getOrderById(Integer id) throws IllegalArgumentException {
+        OrderEntity orderEntityFound = orderRepository.findById(id)
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/service/impl/OrderServiceImpl.java:46]: what if id null? should it be a wrapper Integer?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/LaundryController.java [hunk 6]
```diff
@@ -1,34 +0,0 @@
...
-
-    private LaundryService laundryService;
-
-    @Autowired
-    public LaundryController(OrderService orderService) {
-        this.orderService = orderService;
-    }
-
-    /*@GetMapping
-    public List<LaundryStateEntity> getLaundriesState() {
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/LaundryController.java:26]: ?

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java [hunk 7]
```diff
@@ -1,17 +1,29 @@
...
-@EqualsAndHashCode
 @AllArgsConstructor
+@EqualsAndHashCode
 public class LaundryStateDTO {
 
-    int laundryId;
-    long queueWaitingTime;
-    int version;
-}
+    private int laundryId;
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java]: seems like all the dtos should have the Object fields. Integer and Long

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderDetailEntity.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+@Entity
+@Getter
+@Setter
+@NoArgsConstructor
+@Table(name = "details")
+public class OrderDetailEntity {
+
+    @Id
+    private Integer id;
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderDetailEntity.java]: and vise versa. The entitis should use primitives where the field require not null

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java [hunk 9]
```diff
@@ -1,4 +1,16 @@
...
+import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@AllArgsConstructor
+@EqualsAndHashCode
 public class OrderSubmittedDTO {
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java]: Dto

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderCoordinationDTO.java [hunk 10]
```diff
@@ -12,8 +13,9 @@
 @Setter
 @EqualsAndHashCode
 @AllArgsConstructor
-public class OrderDTO {
```
Comment (andrsuh) [inline taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderCoordinationDTO.java]: Dto

**Review by andrsuh — CHANGES_REQUESTED** [review]

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java (hunk 1 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/outbound/OrderSubmissionDTO.java]: fixed

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/LaundryController.java (hunk 6 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/controllers/LaundryController.java:26]: ? this class is removed

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java (hunk 7 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/LaundryStateDTO.java]: agree
fixed

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderDetailEntity.java (hunk 8 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/entity/OrderDetailEntity.java]: disagree
https://stackoverflow.com/questions/11283709/always-use-primitive-object-wrappers-for-jpa-id-instead-of-primitive-type
https://stackoverflow.com/questions/3535791/primitive-or-wrapper-for-hibernate-primary-keys
https://stackoverflow.com/questions/3535791/primitive-or-wrapper-for-hibernate-primary-keys

File: taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java (hunk 9 above)
Comment (AfanasievYuriy) [reply taskcoordinator/src/main/java/com/microservices/taskcoordinator/dto/inbound/OrderSubmittedDTO.java]: is there any reason at all?

**Review by andrsuh — APPROVED** [review]

**Pull request merged**